haystack_experimental/components/generators/chat/openai.py

```diff
@@ -0,0 +1,155 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from dataclasses import replace
+from typing import Any, Optional, Union
+
+from haystack import component
+from haystack.components.generators.chat.openai import OpenAIChatGenerator as BaseOpenAIChatGenerator
+from haystack.dataclasses import ChatMessage, StreamingCallbackT
+from haystack.tools import Tool, Toolset
+
+from haystack_experimental.utils.hallucination_risk_calculator.dataclasses import HallucinationScoreConfig
+from haystack_experimental.utils.hallucination_risk_calculator.openai_planner import calculate_hallucination_metrics
+
+
+@component
+class OpenAIChatGenerator(BaseOpenAIChatGenerator):
+    @component.output_types(replies=list[ChatMessage])
+    def run(
+        self,
+        messages: list[ChatMessage],
+        streaming_callback: Optional[StreamingCallbackT] = None,
+        generation_kwargs: Optional[dict[str, Any]] = None,
+        *,
+        tools: Optional[Union[list[Tool], Toolset]] = None,
+        tools_strict: Optional[bool] = None,
+        hallucination_score_config: Optional[HallucinationScoreConfig] = None,
+    ) -> dict[str, list[ChatMessage]]:
+        """
+        Invokes chat completion based on the provided messages and generation parameters.
+
+        :param messages:
+            A list of ChatMessage instances representing the input messages.
+        :param streaming_callback:
+            A callback function that is called when a new token is received from the stream.
+        :param generation_kwargs:
+            Additional keyword arguments for text generation. These parameters will
+            override the parameters passed during component initialization.
+            For details on OpenAI API parameters, see [OpenAI documentation](https://platform.openai.com/docs/api-reference/chat/create).
+        :param tools:
+            A list of tools or a Toolset for which the model can prepare calls. If set, it will override the
+            `tools` parameter set during component initialization. This parameter can accept either a list of
+            `Tool` objects or a `Toolset` instance.
+        :param tools_strict:
+            Whether to enable strict schema adherence for tool calls. If set to `True`, the model will follow exactly
+            the schema provided in the `parameters` field of the tool definition, but this may increase latency.
+            If set, it will override the `tools_strict` parameter set during component initialization.
+        :param hallucination_score_config:
+            If provided, the generator will evaluate the hallucination risk of its responses using
+            the OpenAIPlanner and annotate each response with hallucination metrics.
+            This involves generating multiple samples and analyzing their consistency, which may increase
+            latency and cost. Use this option when you need to assess the reliability of the generated content
+            in scenarios where accuracy is critical.
```

Comment:
While the new paper is not published, let's at least link to their previous one, which is related.
```suggestion
            in scenarios where accuracy is critical. For details, see the [research paper](https://arxiv.org/abs/2507.11768).
```